test/higher_order_ops/test_with_effects.py

```diff
@@ -662,6 +680,241 @@ def forward(self, arg0_1, arg1_1):
+            lib.define("foo(Tensor x) -> Tensor")
+
+            def foo_impl(a):
+                return a.clone()
+
+            def foo_bwd(ctx, grad):
+                return torch.ops._mylib.foo(grad)
+
+            for backend in ["CPU", "CUDA", "Meta"]:
+                lib.impl("foo", foo_impl, backend)
+
+            if torch._C._dispatch_has_kernel_for_dispatch_key(
+                "_mylib::foo", "Autograd"
+            ):
+                self.skipTest(
+                    "Double registration of Autograd kernel for test custom op"
+                )
+
+            torch.library.register_autograd("_mylib::foo", foo_bwd)
+
+            from torch._higher_order_ops.effects import (
+                _deregister_effectful_op,
+                _EffectType,
+                _register_effectful_op,
+            )
+
+            _register_effectful_op(torch.ops._mylib.foo.default, _EffectType.ORDERED)
+            try:
+
+                def fn(x, y):
+                    return torch.ops._mylib.foo(x) + y
+
+                def ins_dense_req_grad():
+                    return (
+                        torch.tensor([1.0, 2.0, 3.0], requires_grad=True),
+                        torch.tensor([4.0, 5.0, 6.0], requires_grad=True),
+                    )
+
+                def ins_sc_req_grad():
+                    return (
+                        TwoTensor(
+                            torch.tensor([1.0, 2.0, 3.0], requires_grad=True),
+                            torch.tensor([4.0, 5.0, 6.0], requires_grad=True),
+                        ),
+                        torch.tensor([4.0, 5.0, 6.0], requires_grad=True),
+                    )
+
+                for i, ins_fn in enumerate([ins_dense_req_grad, ins_sc_req_grad]):
+                    ref_ins = ins_fn()
+
+                    ref_out = fn(*ref_ins)
+                    ref_out.sum().backward()
+
+                    compiled_fn = torch.compile(fn, backend="aot_eager", fullgraph=True)
+                    ins = ins_fn()
+                    out = compiled_fn(*ins)
+                    self.assertEqual(ref_out, out)
+                    out.sum().backward()
+                    self.assertEqual(ref_ins[1].grad, ins[1].grad)
+                    self.assertEqual(ref_ins[0].grad, ins[0].grad)
+
+                    fw_graph, bw_graph = get_fw_bw_graph(fn, ins)
+                    if i == 0:
+                        self.assertExpectedInline(
+                            fw_graph.code.strip(),
+                            """\
+def forward(self, primals_1, primals_2, primals_3):
+    with_effects = torch.ops.higher_order.with_effects(primals_1, torch.ops._mylib.foo.default, primals_2);  primals_1 = primals_2 = None
+    getitem = with_effects[0]
+    getitem_1 = with_effects[1];  with_effects = None
+    add = torch.ops.aten.add.Tensor(getitem_1, primals_3);  getitem_1 = primals_3 = None
+    return (getitem, add, getitem)""",
```

Comment:
It looks like we're threading the token from the forward to the backward. Is there a reason why we didn't make a separate token for the backward? This would avoid some of the problems with `must_be_in_backward`, because the token would need to be in the backward.

Reply:
Ivan and I talked a bit offline about it. I thought that the `must_be_in_backward` changes were kind of necessary anyway, if/when we have any other mutations that can show up in the middle of the backward graph (e.g. from FSDP), that we need to ensure the partitioner does not move to the forward.

We could move the backward tokens into tangents too - it just saves us from having to add the extra wrapper logic at runtime in the backward to pass in extra token inputs, and deal with the corresponding extra indexing logic (although we still need the wrapper at the end of the backward to remove the tokens in either case). Let me know if given the above you still have an preference on "backward tokens as saved activations vs tangents"

Reply:
> This approach is also consistent with our future "pre-dispatch trace the backward and specialize later" plan -- it's possible that we won't see the effect token until we specialize, and at that point it is too late to change the forward graph.

oh that is true, that in "pre-dispatch backward graph" world we will probably want the effect token runtime logic to live entirely in the backward graph / as tangents

Reply:
> Maybe a degenerate case 😛 

Can DTensor issue wait_tensor calls? If so, then not a degenerate case. This is similar to Tensor subclassing issuing intermediate mutations in their torch_dispatch for functional operators, we put functionalization underneath subclasses because we were worried about that, right?

>  Let me know if given the above you still have an preference on "backward tokens as saved activations vs tangents"

I still prefer the tokens to be in the backward. Ignoring the need to beef up  "must_be_in_backward", putting the tokens in the backward feels like the morally correct solution (the backward graph is technically a separate graph from the forward graph) and is more composable with what we want to do in the future. How difficult would it be to do this / how much additional time would we spend on it?

Reply:
I'll let @IvanKobzarev comment, but you're definitely right that the case I mentioned above is probably not very degenerate (e.g. if you manage to have a fw/bw graph where the only DTensor is coming from tangents, and so unless the tangents are DTensors then there will otherwise be no collectives in the backward).

On the partitioner changes, I could imagine us needing them if/when `set_()` and friends show up in the backward graph? Although I'm not sure if we should add that "fix" pre-emptively or wait for this case to show up in FSDP2 tracing to get a good test case.